File: examples/footguns/09_uninitialized_read/main.c
/*
 * Footgun: Uninitialized Read — C
 *
 * In C, local variables are not zero-initialized unless you explicitly
 * assign them. Reading an uninitialized variable is undefined behavior:
 * on a real program the "garbage" value comes from whatever was on the
 * stack in a previous call frame. Compilers may or may not warn; -Wall
 * catches simple cases but misses many real-world patterns.
 *
 * This bug is subtle because:
 * - It passes compilation without -Wall or with ignored warnings
 * - It often "works" in debug builds (debug allocators zero memory)
 * - It fails intermittently in release builds when the stack pattern changes
 * - Valgrind / MSan (-fsanitize=memory) can detect it; nothing else can
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/* BROKEN: x is never assigned but is used. On most real programs this
 * reads the garbage left by a previous function's stack frame. */
void uninitialized_local(void) {
    int x; /* declared but never assigned */
    /* gcc -Wall: "warning: 'x' is used uninitialized" — often ignored */
    printf("x = %d\n", x); /* UB: any value, including 0 by accident */
}

/* BROKEN: conditional initialization — only one branch assigns sum.
 * If n == 0 the else branch doesn't run, so sum is unread garbage. */
int conditional_init(int n) {
    int sum;
    if (n > 0) {
        sum = 0;
        for (int i = 0; i < n; i++) sum += i;
    }
    /* when n <= 0, sum is never assigned */
    return sum; /* UB when n <= 0 */
}

/* BROKEN: struct field not initialized — the other fields of the struct
 * are left as whatever bytes were on the stack. Happens frequently when
 * new fields are added to a struct and old initialization code is not
 * updated. */
typedef struct { int a; int b; int c; } Triple;

Triple uninitialized_struct(void) {
    Triple t;
    t.a = 1;
    t.b = 2;
    /* t.c is never assigned */
    return t; /* t.c is garbage */
}

int main(void) {
    printf("--- uninitialized local ---\n");
    uninitialized_local();

    printf("--- conditional init (n=0) ---\n");
    printf("sum = %d\n", conditional_init(0)); /* UB */

    printf("--- uninitialized struct field ---\n");
    Triple t = uninitialized_struct();
    printf("t = {%d, %d, %d}\n", t.a, t.b, t.c); /* t.c is garbage */

    return 0;
}

/*
 * Detect: gcc -fsanitize=memory, Valgrind --track-origins=yes
 * Fix: always initialize at declaration:
 *   int x = 0;
 *   Triple t = {0};       -- zero-initializes all fields
 *   Triple t = {1, 2, 3}; -- explicit field values
 */

File: examples/footguns/09_uninitialized_read/main.go
// Footgun: Uninitialized Read — Go
//
// Go zero-initializes every variable at declaration. There is no such
// thing as an uninitialized variable in Go: `var x int` is always 0,
// `var s string` is always "", `var p *T` is always nil. This is a
// language-level guarantee, not a runtime convention.
//
// The entire category of "uninitialized read" bugs does not exist in Go.
// You cannot declare a variable without a value; the zero value is always
// well-defined for every type.

package main

import "fmt"

// Always safe: x is 0 by language guarantee.
func zeroInitLocal() {
	var x int // always 0 — guaranteed by the spec
	fmt.Printf("x = %d\n", x)
}

// Always safe: sum is 0 before the loop (or stays 0 if n <= 0).
func conditionalInit(n int) int {
	var sum int // zero — no conditional uninitialized path
	for i := 0; i < n; i++ {
		sum += i
	}
	return sum // always valid
}

// Always safe: all struct fields are zero-initialized.
type Triple struct{ A, B, C int }

func zeroInitStruct() Triple {
	var t Triple // {A:0, B:0, C:0} — guaranteed
	t.A = 1
	t.B = 2
	// t.C is 0, not garbage
	return t
}

func main() {
	fmt.Println("--- zero-initialized local ---")
	zeroInitLocal()

	fmt.Println("--- conditional init (n=0) ---")
	fmt.Printf("sum = %d\n", conditionalInit(0)) // 0, not garbage

	fmt.Println("--- zero-initialized struct field ---")
	t := zeroInitStruct()
	fmt.Printf("t = {%d, %d, %d}\n", t.A, t.B, t.C) // t.C == 0
}

// Key insight: Go's zero-value guarantee is not just a convention —
// it is part of the language specification. The runtime enforces it for
// all allocations (stack frames, heap, global variables, channel receives).
// New fields added to a struct are automatically zero-initialized; the C
// "forgot to initialize the new field" bug cannot happen.
//
// The tradeoff: you must be careful that the zero value is meaningful for
// your type (e.g., a mutex's zero value is an unlocked mutex — intentional).
